File: java/src/test/java/cpaint/GivenScenario.java
package cpaint;

import org.junit.jupiter.api.Test;

import static cpaint.DSLAcceptanceTests.*;

class GivenScenario {
    @Test
    void draw_a_canvas_a_line_a_rectangle_and_quit_the_program() throws Exception {
        check(withInput(
                "C 20 5",
                "L 1 3 7 3",
                "L 7 1 7 3",
                "R 15 2 20 5",
                "Q"),

                outputIs("enter command: ",
                        "----------------------",
                        "|                    |",
                        "|                    |",
                        "|                    |",
                        "|                    |",
                        "|                    |",
                        "----------------------",
                        "enter command: ",
                        "----------------------",
                        "|                    |",
                        "|                    |",
                        "|xxxxxxx             |",
                        "|                    |",
                        "|                    |",
                        "----------------------",
                        "enter command: ",
                        "----------------------",
                        "|      x             |",
                        "|      x             |",
                        "|xxxxxxx             |",
                        "|                    |",
                        "|                    |",
                        "----------------------",
                        "enter command: ",
                        "----------------------",
                        "|      x             |",
                        "|      x       xxxxxx|",
                        "|xxxxxxx       x    x|",
                        "|              x    x|",
                        "|              xxxxxx|",
                        "----------------------",
                        "enter command: "));
    }
}

File: scala/src/test/scala/cpaint/GivenScenario.scala
package cpaint

import org.apache.tools.ant.filters.StringInputStream
import org.scalatest.{FunSuite, Matchers}

class GivenScenario extends FunSuite with Matchers {

  test("draw a canvas") {
    check(
      withInput(
        "C 20 5"),

      outputIs("enter command: ",
        "----------------------",
        "|                    |",
        "|                    |",
        "|                    |",
        "|                    |",
        "|                    |",
        "----------------------",
        "enter command: "))
  }

  private def check(input: StringInputStream, expectedOutput: String): Any = {
    val out = new java.io.ByteArrayOutputStream()
    Console.withIn(input) {
      Console.withOut(out) {
        new ConsolePaint().execute()
      }
    }

    out.toString() should be(expectedOutput)
  }

  private def withInput(inputs: String*) = {
    new StringInputStream(joinWithLineSeparator(inputs))
  }

  private def outputIs(expectedOutput: String*) = {
    joinWithLineSeparator(expectedOutput)
  }

  private def joinWithLineSeparator(input: Seq[String]) = {
    input.mkString(System.lineSeparator())
  }
}
